### ISSUE TITLE
[Feature Request] Soft Patch roms (non-cd)

### ISSUE THREAD

**Issue by Paul-Colucci** [libretro/Genesis-Plus-GX#211]
It would be great to be able to soft patch non-cd roms, or at least expose an option so you can soft patch at your own risk.

**Comment by hizzlekizzle** [comment]
While such a thing may be possible for some cores, as I recall this core needs the actual location of the file regardless of what type of media is being loaded.

**Comment by ekeeke** [comment]
Well, in theory, since soft-patching is likely only usable for ROM files (not CD image files) and since ROM file name is apparently always available from libretro, it could be possible to modify the "load_archive" function in libretro.c to use the ROM pointer/length provided by libretro API and initialize the ROM buffer required by the emulator, as well as still returning the ROM file extension and ROM size expected from this function. Note that this function is only called by the emulator if the loaded file is a ROM file, not if it is a CD image file.

From my understanding, as long as the loaded file path is available from libretro and the emulator can still open and read from the file, it does not matter to the emulator if the file was loaded into RAM by the frontend.

The only problem seems  that libretro requires the core to set "need_fullpath" setting to FALSE to allow soft-patching but I don't know what will happen if this set to FALSE by default and the user tries to load a CD image file on systems that don't have enough RAM to hold the entire file.

If the frontend behavior is to to return a load error in this case then this means this will break CD support on any platform with limited RAM ressources.

**Comment by Paul-Colucci** [comment]
It would be really nice to have.  There are a lot of people who prefer soft patching to preserve the integrity of the original rom, and not having to fork rom sets

**Comment by ekeeke** [comment]
I think everybody understand why soft-patching is interesting and desired, it's not the issue here.

As explained above, the required change in genesis plus gx libretro portion to use patched ROM data instead of reading from ROM file are minimal and quite simple (I will try to push a commit in my repository to show the needed changes, even if it will have no effect without retroarch being fixed as well). The problem is that the frontend (rcetroarch) ONLY patches ROM data if the core asks him to ALWAYS load entire file content into RAM and it's not possible to tell the frontend to "do this only for ROMs, not CD image files" as the frontend has no way to distinguish between those, so whenever you will try to load a CD game, Retroarch will fail loading it unless you have enough RAM (Sega CD games can be up 600 MB).

I also tried solution 1 above (to make it user configurable) but it does not work as core settings are only available once a game is loaded.

So, unless I'm missing something (NB: this is a call to libretro devs to correct me if I'm wrong) the only solution is to also modify the frontend (retroarch) using either solution 2 or 3 above.

**Comment by hizzlekizzle** [comment]
Yeah, I believe you're correct. #3 indeed seems most doable, just at a glance.

**Comment by ekeeke** [comment]
I have submitted a commit in my repository (https://github.com/ekeeke/Genesis-Plus-GX/commit/81becffb6c9d1ee9c754439eb4242633a5ae9076) with the modifications required to support ROM files already loaded in memory by Retroarch.

For the moment, since the core still sets "need_fullpath" to TRUE, Retroarch never attempts to load ROM file into memory and disable auto-patching so provided info->data is always NULL.

Therefore, it still requires frontend modification (either solution 2 or 3 above) to be fully functional.

NB: I quickly tested it worked by compiling a build with "need_fullpath" set to False and verified the loaded file was indeed loaded into memory. As already explained, this solution is not viable as it requires up to 600 MB of RAM for Mega CD games, and thus would make the core unusable for Mega CD emulation on many platforms supported by Retroarch (and even on platform with enough RAM, forcing such high RAM use is not recommended practice unless explicitly enabled by the user).

**Comment by ichee** [comment]
Adding an option in the libretro core to preload game into memory would solve this problem by linking soft patching to that option. RAM usage shouldn't be a concern then as long as the option is disabled by default.

**Comment by ekeeke** [comment]
As explained above, adding a core option is not possible as, by design in retroarch, core settings are not accessible until game file has been loaded so it's already too late.

As explained in linked pull request, the solution requires adding an extension to libretro API and some retroarch modification to allow cores to enable soft-patching for some files (based on file extension) even when need_fullpath is set to TRUE by the core.

**Comment by ichee** [comment]
In the case of 64-bit builds, at least for Windows and Linux, it can just be loaded into memory regardless of extension since amount of memory isn't going to be a concern for those platforms since the average user there has at least 8GB or RAM, so regardless of whether it's a rom or CD, there should be no issues, need_fullpath can then just be set to false.

If these are still a concern for you, a temporary solution could be to provide a build with the above mentioned here or submitted to the Retroarch team as an experimental build until a proper agreed-upon solution can be found for all.

**Comment by ekeeke** [comment]
You are free to build your own core version with need_fullpath set to false if you cannot wait for a proper solution on retroarch side but I would definitively not release such build myself or force this for windows users assuming they have enough free RAM.

**Comment by ichee** [comment]
That's why I suggested a build posted here or released explicitly as an experimental build, not to replace the current master builds.
It would also help to get additional feedback testing in case there are other unforeseen issues.
I think most users would appreciate the thought given the lack of soft-patching support for this emulator, and given the fact that any necessary changes on the Retroarch side are not likely to happen any time soon.

**Comment by ichee** [comment]
Building with need_fullpath set to false does not load game, rom or CD, into memory, nor does soft-patching work with your recent changes to master.
There's more going on here than a simple flag change during compilaton, whether that's more significant changes needed to your core, or Retroarch completely ignoring the need_fullpath flag state.

**Comment by ichee** [comment]
I managed to get the build working with need_fullpath set to 'false' and softpatching is now working.

Just a for your info, the reason it didn't work before was because the makefile needed to be cleaned first, despite being a clean install and fresh build environment.

For anyone that's interested, here is the build with softpatching functional, Windows 64-bit.

(note that all games, roms and cd images alike, will be loaded into memory, so make sure you have at least 4GB, more for multi-cd games loaded with a .m3u file)

Core based off of ekeeke's latest master [here](https://github.com/ekeeke/Genesis-Plus-GX).

[genesis_plus_gx_libretro.zip](https://github.com/libretro/Genesis-Plus-GX/files/5146146/genesis_plus_gx_libretro.zip)
